### PR TITLE
Fix/single product btn purchasable in stock

### LINKED ISSUE #251: Amazon Pay Button is displayed even when Out Of Stock
**Describe the bug**
https://wordpress.org/support/topic/button-is-displayed-even-when-out-of-stock/

Amazon Pay button is displayed on Product pages even when product is Out Of Stock. 
**To Reproduce**
1. on the Settings activate the option "Amazon Pay on product pages"
2. On Fron-end go to a product page that is out of stock
3. Button still displayed

**Screenshots**
image

**Expected behavior**
Button should not be displayed when product is out of stock
**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only the extension is active.
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by dpanta94**
### All Submissions:

* [x] Does your code follow the [Extendables](https://extendomattic.wordpress.com/standardizations/) standards?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changes proposed in this Pull Request:

Closes #251 .

### How to test the changes in this Pull Request:

1. Enable the Amazon Pay for single product pages from the plugin's options
2. Set a product to in stock or backorder. The Amazon pay button should appear.
3. Set the product as out of stock. The product should dissappear and Amazon pay related scripts should not be loaded in the page.

### Other information:

* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changelog entry

> Fix - Amazon pay button should appear on single product page only when the product is purchasable and in stock.

### CODE REVIEW

File: includes/class-wc-gateway-amazon-payments-advanced.php [hunk 1]
```diff
@@ -2815,13 +2827,32 @@ public function load_scripts_globally_if_button_enabled_on_cart( $load_scripts )
...
-		return $load_scripts ? $load_scripts : is_product() && $this->is_product_button_enabled();
+		if ( ! is_product() ) {
+			return $load_scripts;
+		}
+
+		if ( ! $this->is_product_button_enabled() ) {
+			return $load_scripts;
+		}
+
+		$product = wc_get_product();
```
Comment (kalessil) [inline includes/class-wc-gateway-amazon-payments-advanced.php:2845]: Those 2 ifs are structurally code duplicates of what's we are changing in 'maybe_separator_and_checkout_button_single_product'.  Can we refactor that?

File: tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php [hunk 2]
```diff
@@ -136,4 +123,118 @@ public function test_classic_process_payment_succeeds() : void {
...
+	public function test_maybe_separator_and_checkout_button_single_product() : void {
+		ob_start();
+		self::$gateway->maybe_separator_and_checkout_button_single_product();
+		$should_fail_because_no_global_product = ob_get_clean();
+
+		$this->assertEmpty( $should_fail_because_no_global_product );
+
+		$test_product = WC_Helper_Product::create_and_optionally_save_simple_product( true );
+
+		$test_product->set_stock_status( 'outofstock' );
```
Comment (kalessil) [inline tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php:141]: For readability purposes, can we drop the empty lines before and after this line?

File: tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php [hunk 3]
```diff
@@ -136,4 +123,118 @@ public function test_classic_process_payment_succeeds() : void {
...
+
+		$GLOBALS['post'] = get_post( $test_product->get_id() );
+
+		ob_start();
+		self::$gateway->maybe_separator_and_checkout_button_single_product();
+		$should_fail_because_out_of_stock = ob_get_clean();
+
+		$this->assertEmpty( $should_fail_because_out_of_stock );
+
+		$test_product->set_stock_status( 'instock' );
```
Comment (kalessil) [inline tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php:153]: For readability purposes, can we drop the empty lines before and after this line?

File: tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php [hunk 4]
```diff
@@ -136,4 +123,118 @@ public function test_classic_process_payment_succeeds() : void {
 			$mock_gateway->process_payment( $order->get_id() )
 		);
 	}
+
+	/**
+	 * Test maybe_separator_and_checkout_button_single_product method.
+	 *
+	 * @return void
+	 */
+	public function test_maybe_separator_and_checkout_button_single_product() : void {
```
Comment (kalessil) [inline tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php:132]: The test case is a bit complicated. IIUC, we can split it 2:
- one test case covering non-product context
- one test with a product, but also a data provider providing stock statuses

I might miss something, but each test case covers one scenario.

File: tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php [hunk 5]
```diff
@@ -136,4 +123,118 @@ public function test_classic_process_payment_succeeds() : void {
...
+	public function test_load_scripts_on_product_pages() {
+		// If given true it should always return true, no matter what.
+		$this->assertTrue( self::$gateway->load_scripts_on_product_pages( true ) );
+
+		// Should return false, since this is not a product page.
+		$this->assertFalse( self::$gateway->load_scripts_on_product_pages( false ) );
+
+		$test_product = WC_Helper_Product::create_and_optionally_save_simple_product( true );
+
+		$test_product->set_stock_status( 'outofstock' );
```
Comment (kalessil) [inline tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php:192]: For readability purposes, can we drop the empty lines before and after this line?

File: tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php [hunk 6]
```diff
@@ -136,4 +123,118 @@ public function test_classic_process_payment_succeeds() : void {
...
+		$GLOBALS['post'] = get_post( $test_product->get_id() );
+
+		// Force the global query to be a singular product query.
+		$GLOBALS['wp_query']->is_singular    = true;
+		$GLOBALS['wp_query']->queried_object = $GLOBALS['post'];
+
+		// Scripts should not be loaded since product is out of stock.
+		$this->assertFalse( self::$gateway->load_scripts_on_product_pages( false ) );
+
+		$test_product->set_stock_status( 'instock' );
```
Comment (kalessil) [inline tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php:206]: For readability purposes, can we drop the empty lines before and after this line?

**Review by kalessil — APPROVED** [review]
The fix is straightforward, but we can simplify it, + we need to iterate on tests implementation and make them more readable (and maintainable) by future us.

**Review by kalessil — CHANGES_REQUESTED** [review]
Sorry, I hit the wrong button.

File: includes/class-wc-gateway-amazon-payments-advanced.php (hunk 1 above)
Comment (dpanta94) [reply includes/class-wc-gateway-amazon-payments-advanced.php]: Yea makes sense. Refactored!

File: tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php (hunk 4 above)
Comment (kalessil) [reply tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php:132]: I wish I had more capacity to spend here, but we are good at having the tests.
My comments imply a bit of pair programming, to improve the test structure a bit - let's ignore my comment.

**Review by kalessil — APPROVED** [review]
LGTM!

**Pull request merged**